src/scripts/data_file_processing/healthsystem/equipment/equipment_availability_estimation.py

```diff
@@ -0,0 +1,273 @@
+levels_4_and_5_lookup = mfl.loc[mfl['Facility_Level'].isin(['4', '5'])].set_index('Facility_Level')['Facility_ID'].to_dict()
+mask_to_levels_4_and_5 = final_equipment_availability_export['Facility_Level'].isin(['4', '5'])
+final_equipment_availability_export.loc[mask_to_levels_4_and_5, 'Facility_ID'] = \
+    final_equipment_availability_export.loc[mask_to_levels_4_and_5, 'Facility_Level'].map(levels_4_and_5_lookup)
+
+assert final_equipment_availability_export['Facility_ID'].notnull().all()
+final_equipment_availability_export['Facility_ID'] = final_equipment_availability_export['Facility_ID'].astype(int)
+
+# - Format as a pd.Series with multi-index (Facility_ID, Item_Code)
+final_equipment_availability_export = final_equipment_availability_export.set_index(['Facility_ID', 'Item_Code'])['Pr_Available']
+
+# Extrapolate so that there is an estimated availability for every item at every level
+# - Create "full" dataset, where we force that there is probability of availability for every item_code at every
+#  observed facility
+_all_fac_ids = mfl.Facility_ID.unique()
+_all_item_codes = pd.read_csv(
+    resourcefilepath /'healthsystem/infrastructure_and_equipment/ResourceFile_EquipmentCatalogue.csv'
+)['Item_Code'].unique()
+final_equipment_availability_export_full = pd.Series(
+    index=pd.MultiIndex.from_product(
+        [_all_fac_ids, _all_item_codes],
+        names=["Facility_ID", "Item_Code"]
+    ),
+    data=float("nan"),
+    name='Pr_Available'
+).combine_first(final_equipment_availability_export)
+
+print(f''
+      f'Fraction of missing data when requiring data for all items at all facilities = '
+      f'{round(100 * final_equipment_availability_export_full.isnull().sum() / len(final_equipment_availability_export_full))}'
+      f'%')
+
+# Impute availability of missing item_codes in a facility_id... from the mean availability of the same items in other facilities of the same level
+facility_levels = final_equipment_availability_export_full.index.get_level_values('Facility_ID').map(
+    mfl.set_index('Facility_ID')['Facility_Level'].to_dict()
+)
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby(['Item_Code', facility_levels]).transform(lambda x: x.fillna(x.mean()))
+
+# Remaining misising data include those item_codes that are never seen anywhere: interpolate these from the average availability of all other items, stratified by level
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby([facility_levels]).transform(lambda x: x.fillna(x.mean()))
+
+# Remaining missing data are for items_codes in facility_ids for which there is no information at all in the facility_level
+# Impute the availability for these items in facilities as ... the average of other items available at other facilities
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby("Item_Code").transform(lambda x: x.fillna(x.mean()))
```

Comment:
This is my proposal for doing the extrapolation to item_codes/ facility_id that are not directly observed in the data.
Please let me know if it looks ok, or you see anything suspicious in the resulting data....?

Reply:
This looks good. The only thing I'm concerned about is interpolating the value of cheaper items like thermometer from expensive items like X-rays which are likely to have a lower level of availability. Because equipment costing in the HSSP-III differentiates between equipment which cost <$1000 and those which cost more, we could use the same distincition to interpolate availability. Let me try to add this change.

Reply:
Very good point! :)

If it becomes complicated to do the interpolation using these simples rules when using multiple criteria, we could always estimate a linear model for the availability, and use that to fill in the blanks;

e.g.,
availability ~ item_code + facility_level + district + price_is_high

Reply:
Thanks, Tim. I've pushed the update now. I just updated the interpolation method to include (binary) equipment category and facility levels (instead of just facility levels), as it seemed more straightforward that using a linear model.